lib/ClusterShell/CLI/Clush.py

```diff
@@ -966,7 +966,8 @@ def main():
     task.set_default("USER_handle_SIGUSR1", user_interaction)
 
     task.excepthook = sys.excepthook
-    task.set_default("USER_stdin_worker", not (sys.stdin.isatty() or \
+    task.set_default("USER_stdin_worker", not ((sys.stdin is not None and \
```

Comment:
The original reporter reported that this patch does not fix the issue. There's another access of sys.stdin in an area of the code that I thought shouldn't be reached. Looking at this code, I've either confused myself, or the original code was incorrect.  If stdin is a tty, it was setting this to false - is that the correct behavior?

Reply:
There are many cases here that is a bit confusing. Let me explain some of the possible conditions:

TTY stdin: If stdin is connected to a tty, `USER_stdin_worker` is set to False, and `bind_stdin()` is not called, but user_interaction might be set to True (we also check for stdou) to enable terminal interaction (handled by `ttyloop()`).


non-TTY stdin: `bind_stdin()`  is called when a (valid) non-tty stdin is used and also when `--nostdin` is not specified. It creates a thread that is dedicated to read stdin and push the data to a "task port" using a handler `StdInputHandler()`.

the default "USER_stdin_worker" controls whether bind_stdin() is called or not.

This is used for example when we do `echo "foo" | clush -w ... cmd`

If `sys.stdin` is None, `USER_stdin_worker` should be set to False so that bind_stdin is not called. Your patch set `USER_stdin_worker` to True if sys.stdin is None. We want the opposite.

Reply:
It makes sense to me now. Thanks for the explanation.